README.md
# Little Wire Python library

<p align="center">
  <a href="https://github.com/tiacsys/verylittlewire/actions?query=workflow%3ACI">
    <img src="https://img.shields.io/github/workflow/status/tiacsys/verylittlewire/CI/main?label=CI&logo=github&style=flat-square" alt="CI Status" >
  </a>
  <a href="https://verylittlewire.readthedocs.io">
    <img src="https://img.shields.io/readthedocs/verylittlewire.svg?logo=read-the-docs&logoColor=fff&style=flat-square" alt="Documentation Status">
  </a>
  <a href="https://codecov.io/gh/tiacsys/verylittlewire">
    <img src="https://img.shields.io/codecov/c/github/tiacsys/verylittlewire.svg?logo=codecov&logoColor=fff&style=flat-square" alt="Test coverage percentage">
  </a>
</p>
<p align="center">
  <a href="https://python-poetry.org/">
    <img src="https://img.shields.io/badge/packaging-poetry-299bd7?style=flat-square&logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAA4AAAASCAYAAABrXO8xAAAACXBIWXMAAAsTAAALEwEAmpwYAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAJJSURBVHgBfZLPa1NBEMe/s7tNXoxW1KJQKaUHkXhQvHgW6UHQQ09CBS/6V3hKc/AP8CqCrUcpmop3Cx48eDB4yEECjVQrlZb80CRN8t6OM/teagVxYZi38+Yz853dJbzoMV3MM8cJUcLMSUKIE8AzQ2PieZzFxEJOHMOgMQQ+dUgSAckNXhapU/NMhDSWLs1B24A8sO1xrN4NECkcAC9ASkiIJc6k5TRiUDPhnyMMdhKc+Zx19l6SgyeW76BEONY9exVQMzKExGKwwPsCzza7KGSSWRWEQhyEaDXp6ZHEr416ygbiKYOd7TEWvvcQIeusHYMJGhTwF9y7sGnSwaWyFAiyoxzqW0PM/RjghPxF2pWReAowTEXnDh0xgcLs8l2YQmOrj3N7ByiqEoH0cARs4u78WgAVkoEDIDoOi3AkcLOHU60RIg5wC4ZuTC7FaHKQm8Hq1fQuSOBvX/sodmNJSB5geaF5CPIkUeecdMxieoRO5jz9bheL6/tXjrwCyX/UYBUcjCaWHljx1xiX6z9xEjkYAzbGVnB8pvLmyXm9ep+W8CmsSHQQY77Zx1zboxAV0w7ybMhQmfqdmmw3nEp1I0Z+FGO6M8LZdoyZnuzzBdjISicKRnpxzI9fPb+0oYXsNdyi+d3h9bm9MWYHFtPeIZfLwzmFDKy1ai3p+PDls1Llz4yyFpferxjnyjJDSEy9CaCx5m2cJPerq6Xm34eTrZt3PqxYO1XOwDYZrFlH1fWnpU38Y9HRze3lj0vOujZcXKuuXm3jP+s3KbZVra7y2EAAAAAASUVORK5CYII=" alt="Poetry">
  </a>
  <a href="https://github.com/ambv/black">
    <img src="https://img.shields.io/badge/code%20style-black-000000.svg?style=flat-square" alt="black">
  </a>
  <a href="https://github.com/pre-commit/pre-commit">
    <img src="https://img.shields.io/badge/pre--commit-enabled-brightgreen?logo=pre-commit&logoColor=white&style=flat-square" alt="pre-commit">
  </a>
</p>
<p align="center">
  <a href="https://pypi.org/project/verylittlewire/">
    <img src="https://img.shields.io/pypi/v/verylittlewire.svg?logo=python&logoColor=fff&style=flat-square" alt="PyPI Version">
  </a>
  <img src="https://img.shields.io/pypi/pyversions/verylittlewire.svg?style=flat-square&logo=python&amp;logoColor=fff" alt="Supported Python versions">
  <img src="https://img.shields.io/pypi/l/verylittlewire.svg?style=flat-square" alt="License">
</p>

The Little Wire Python library provides an interface to the Little Wire USB Multi-Tool developed by Ihsan Kehribar.

## Installation

Install this via pip (or your favourite package manager):

`pip install verylittlewire`

## Contributors ✨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- prettier-ignore-start -->
<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tbody>
    <tr>
      <td align="center" valign="top" width="14.28%"><a href="http://www.li-pro.net/stephan-linz.phtml"><img src="https://avatars.githubusercontent.com/u/5664060?v=4?s=80" width="80px;" alt="Stephan Linz"/><br /><sub><b>Stephan Linz</b></sub></a><br /><a href="https://github.com/tiacsys/verylittlewire/commits?author=rexut" title="Code">💻</a> <a href="#ideas-rexut" title="Ideas, Planning, & Feedback">🤔</a> <a href="https://github.com/tiacsys/verylittlewire/commits?author=rexut" title="Documentation">📖</a></td>
      <td align="center" valign="top" width="14.28%"><a href="https://github.com/adajoh99"><img src="https://avatars.githubusercontent.com/u/1729678?v=4?s=80" width="80px;" alt="Adam Johnson"/><br /><sub><b>Adam Johnson</b></sub></a><br /><a href="https://github.com/tiacsys/verylittlewire/commits?author=adajoh99" title="Code">💻</a></td>
      <td align="center" valign="top" width="14.28%"><a href="https://soundcloud.com/ihsankehribar/sets/bits-and-pieces"><img src="https://avatars.githubusercontent.com/u/360652?v=4?s=80" width="80px;" alt="Ihsan Kehribar"/><br /><sub><b>Ihsan Kehribar</b></sub></a><br /><a href="#ideas-kehribar" title="Ideas, Planning, & Feedback">🤔</a></td>
      <td align="center" valign="top" width="14.28%"><a href="http://omer.kilic.name/"><img src="https://avatars.githubusercontent.com/u/9847?v=4?s=80" width="80px;" alt="Omer Kilic"/><br /><sub><b>Omer Kilic</b></sub></a><br /><a href="#ideas-omerk" title="Ideas, Planning, & Feedback">🤔</a></td>
    </tr>
  </tbody>
</table>

<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->

<!-- ALL-CONTRIBUTORS-LIST:END -->
<!-- prettier-ignore-end -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!

## Credits

This package was created with
[Cookiecutter](https://github.com/cookiecutter/cookiecutter) and the
[browniebroke/cookiecutter-pypackage](https://github.com/browniebroke/cookiecutter-pypackage)
project template.


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.4.1 (2022-11-25)
### Fix
* **feat:** Typos in comment strings ([`ee572a0`](https://github.com/tiacsys/verylittlewire/commit/ee572a055005b181602f248936ea8c5e3694ba12))
* **test:** Typos in comment strings ([`d81cb8f`](https://github.com/tiacsys/verylittlewire/commit/d81cb8f57e0ff60478dd3b341822df3ed703915a))
* **docs:** Typos in contributing rules ([`2ef7bd2`](https://github.com/tiacsys/verylittlewire/commit/2ef7bd2d611d067629b8141286a7a22fab7d29dd))
* **docs:** Correct link to cookiecutter ([`cece7dc`](https://github.com/tiacsys/verylittlewire/commit/cece7dc4a3bd8e6502367409c440c0df5cdbfb0c))
* **docs:** Remove auto-generated files when clean ([`c10ed0e`](https://github.com/tiacsys/verylittlewire/commit/c10ed0e8514d6c6178a642be4d68d66dc9bd0d36))
* **docs:** Autosummary templates for class methods ([`5698b98`](https://github.com/tiacsys/verylittlewire/commit/5698b986c97025a8baa7c98e6adde4ea26b68ea8))
* Remove mypy ignoring union-attr check ([`de7907c`](https://github.com/tiacsys/verylittlewire/commit/de7907c1e1e404ed7ea323f2c37e71eb8dcb2776))

### Documentation
* **api:** New docstring: Little Wire PWM ([`6a3ca19`](https://github.com/tiacsys/verylittlewire/commit/6a3ca19d78ac11b7d0e103595036fd705772b3c6))
* **api:** New docstring: Little Wire ADC ([`3c5145f`](https://github.com/tiacsys/verylittlewire/commit/3c5145fecb12f8d60c29d0b81a91e72cab5185b1))
* **api:** New docstring: Little Wire GPIO ([`337cd2a`](https://github.com/tiacsys/verylittlewire/commit/337cd2aa06bb235098997dcd524f0cd62f9328b5))
* **api:** New docstring: Device.readFirmwareVersion ([`29eef91`](https://github.com/tiacsys/verylittlewire/commit/29eef913581f5588b4700d108bd6684e52c55e29))
* **api:** Add USB Control Transfer Requests ([`a4a4d5e`](https://github.com/tiacsys/verylittlewire/commit/a4a4d5e477ddeba3c1e0a366d80f6c8fa8e67150))
* **api:** New docstring: Device.readSerialNumber ([`c8960fe`](https://github.com/tiacsys/verylittlewire/commit/c8960fefbc69372523b2f6145fbf3e99d4f2ef85))
* **api:** New docstring: Device.__init__ ([`0035d0b`](https://github.com/tiacsys/verylittlewire/commit/0035d0b9d99d211f46b62c72fad7f016179ff8c9))
* **api:** New docstring: Device.lw ([`b874819`](https://github.com/tiacsys/verylittlewire/commit/b874819094e5aab66ce56f94753ff270f51bbc68))
* **api:** New docstring: PRODUCT_ID ([`d586c59`](https://github.com/tiacsys/verylittlewire/commit/d586c59f55509297eeeddb9df5026ef755113c83))
* **api:** New docstring: VENDOR_ID ([`e589e81`](https://github.com/tiacsys/verylittlewire/commit/e589e811c141acc7eee6f1c808a687440e2960c9))
* Precise device module description ([`b10d39f`](https://github.com/tiacsys/verylittlewire/commit/b10d39fcc43dce11d3df8be9bb9b62f9393e3f11))
* Provide a project spelling word list ([`8f06301`](https://github.com/tiacsys/verylittlewire/commit/8f06301311b6f03c068157c586a274276577c6f0))
* More rules to use Sphinx ([`fbe2059`](https://github.com/tiacsys/verylittlewire/commit/fbe2059aa877f9d4276ef17318b76b20b3bb2da1))
* **api:** Add API reference with autosummary ([`dabe0bd`](https://github.com/tiacsys/verylittlewire/commit/dabe0bdf2dce78a7b476a53b14c741bc20fd6183))
* **usage:** Add simple doctest as example ([`29a7c63`](https://github.com/tiacsys/verylittlewire/commit/29a7c63114f077d19127ea040c10e47ffc76d1c9))
* Enable LinuxDoc flat-table ([`d0bb933`](https://github.com/tiacsys/verylittlewire/commit/d0bb9332b57936f07779d4ab9db21e163625178a))
* Enable external links extensions ([`263503e`](https://github.com/tiacsys/verylittlewire/commit/263503effc4e627a2b51eeff665ce169cad7e30e))
* Fine tuning for Sphinx link checker ([`d7c46f6`](https://github.com/tiacsys/verylittlewire/commit/d7c46f67e882c864f5f4dd67c92f524ca2adc5d9))
* Enable spell checker ([`6a13be1`](https://github.com/tiacsys/verylittlewire/commit/6a13be1df59fc3ecabc222ab8f6bc88eee00670f))
* Enable auto-documentation extensions ([`26b6e30`](https://github.com/tiacsys/verylittlewire/commit/26b6e30a028d424a4a016f2f33d1ea73aed99f79))
* Enable auto-section labeling extension ([`277cb91`](https://github.com/tiacsys/verylittlewire/commit/277cb9144d4f977b040b6fe383d2f7b2f6dd2eed))
* Change copyright to TiaC Systems ([`7bd4833`](https://github.com/tiacsys/verylittlewire/commit/7bd4833f5f42d1e3bcf615bfee7cf9c27b467cc2))

## v0.4.0 (2022-11-25)
### Feature
* **example:** Add Little Wire LED loop back fade ([`6be490f`](https://github.com/tiacsys/verylittlewire/commit/6be490fc58f9015c1bc218467a2d9df15e18a040))
* **example:** Add Little Wire LED fade on PIN1 ([`8bc7938`](https://github.com/tiacsys/verylittlewire/commit/8bc7938e82e81466102193470f2b984b7fff6e64))
* Add Little Wire PWM prescaler value setup ([`a13d78c`](https://github.com/tiacsys/verylittlewire/commit/a13d78cf68f0c70c30bca5543945f928295c23ff))
* Add Little Wire PWM compare value setup ([`ed40e74`](https://github.com/tiacsys/verylittlewire/commit/ed40e7499070e7659f5c3cd135f598a89d41f891))
* Add Little Wire PWM system wide stop ([`793ac25`](https://github.com/tiacsys/verylittlewire/commit/793ac2545270aba2669347ff6e8851dd95dbddc7))
* Add Little Wire PWM system wide init ([`e078a6e`](https://github.com/tiacsys/verylittlewire/commit/e078a6e99eb256bc348520be1b0a16c455475114))
* Add Little Wire PWM pin (channel) names ([`4470935`](https://github.com/tiacsys/verylittlewire/commit/4470935b324a348bdc9335ddb8084455b326d59c))

## v0.3.0 (2022-11-25)
### Feature
* **example:** Add Little Wire analog in on PIN2/3 ([`839aeee`](https://github.com/tiacsys/verylittlewire/commit/839aeeed17c62ec83aedc8199f92a337d593f8f1))
* Add Little Wire ADC analog read in ([`976355d`](https://github.com/tiacsys/verylittlewire/commit/976355d3edc458c38c2c691955312b1181680e87))
* Add Little Wire ADC vref setup ([`ee84c84`](https://github.com/tiacsys/verylittlewire/commit/ee84c8437570f610aaff2e7a6b76ad11cd889b53))

## v0.2.0 (2022-11-25)
### Feature

tests/__init__.py
"""Unit test package for verylittlewire."""


tests/test_main.py
from verylittlewire.main import add


def test_add():
    assert add(1, 1) == 2


src/verylittlewire/device.py
# SPDX-License-Identifier: MIT
#
# Copyright (c) 2022 Stephan Linz <linz@li-pro.net>
# Copyright (c) 2012 Adam Johnson <apjohnson@gmail.com>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy of
# this software and associated documentation files (the "Software"), to deal in
# the Software without restriction, including without limitation the rights to
# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
# of the Software, and to permit persons to whom the Software is furnished to do
# so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

"""
Very Little Wire device module.

The module provides a Python interface to the Little Wire USB Multi-Tool
developed by Ihsan Kehribar.

This file is a direct Python translation of the C/C++ library developed by Ihsan
Kehribar <ihsan@kehribar.me> and Omer Kilic <omerkilic@gmail.com>. The C/C++
library was released under the same license. Adam Johnson did the translation
based on the C/C++ library version 0.9 for Python 2. This new reimplementation
is an effort to port the Python 2 translation to Python 3 and adapt the changes
in the C/C++ library from version 0.9 to the latest version 1.2.

:license: MIT, see LICENSE for details.
"""

import usb

VENDOR_ID = 0x1781
PRODUCT_ID = 0x0C9F
USB_TIMEOUT = 5000

# Little Wire pin modes and states
INPUT = 1
OUTPUT = 0

HIGH = 1
LOW = 0

# GPIO pin enumeration
PIN1 = 1
PIN2 = 2
PIN3 = 5
PIN4 = 0


class Device:
    """
    Class to control a LittleWire USB Multi-Tool.
    """

    lw = None

    def __init__(self) -> None:
        """
        Finds the first Little Wire device and attaches to it.
        """

        self.lw = usb.core.find(idVendor=VENDOR_ID, idProduct=PRODUCT_ID)

        if self.lw is None:
            raise ValueError("Device not Found!")

        self.lw.set_configuration()

    def readSerialNumber(self) -> str:
        """
        Returns Little Wire serial number.
        """

        serial = usb.util.get_string(
            self.lw, self.lw.iSerialNumber  # type: ignore[union-attr]
        )

        return str(serial)

    def readFirmwareVersion(self) -> str:
        """
        Returns Little Wire firmware version.
        """

        result = self.lw.ctrl_transfer(  # type: ignore[union-attr]
            bmRequestType=0xC0,
            bRequest=34,
            wValue=0,
            wIndex=0,
            data_or_wLength=8,
            timeout=USB_TIMEOUT,
        )

        version = result.pop()

        return str((version & 0xF0) >> 4) + "." + str(version & 0x0F)

    def pinMode(self, pin: int, mode: int) -> None:
        """
        Sets GPIO pins to INPUT(1) or OUTPUT(0).
        """

        if mode == INPUT:
            self.lw.ctrl_transfer(  # type: ignore[union-attr]
                bmRequestType=0xC0,
                bRequest=13,
                wValue=pin,
                wIndex=0,
                data_or_wLength=8,
                timeout=USB_TIMEOUT,
            )
        else:
            self.lw.ctrl_transfer(  # type: ignore[union-attr]
                bmRequestType=0xC0,
                bRequest=14,
                wValue=pin,
                wIndex=0,
                data_or_wLength=8,
                timeout=USB_TIMEOUT,
            )

    def digitalWrite(self, pin: int, state: int) -> None:
        """
        Writes a digital HIGH (1) or LOW (0) to the selected GPIO.
        """

        if state == HIGH:
            self.lw.ctrl_transfer(  # type: ignore[union-attr]
                bmRequestType=0xC0,
                bRequest=18,
                wValue=pin,
                wIndex=0,
                data_or_wLength=8,
                timeout=USB_TIMEOUT,
            )
        else:
            self.lw.ctrl_transfer(  # type: ignore[union-attr]
                bmRequestType=0xC0,
                bRequest=19,
                wValue=pin,
                wIndex=0,
                data_or_wLength=8,
                timeout=USB_TIMEOUT,
            )


# vim: tw=80 ts=4 sw=4 sts=4 sta et ai nu


Python/library v1.0/littleWire.py
# File: littleWire.py
# Version: v1.0
# Purpose: Provides a Python interface to the Little Wire USB Multi-Tool
#          developed by Ihsan Kehribar
# Author: Adam Johnson
# Copyright 2012 by Adam Johnson <apjohnson@gmail.com>
#
# This file is a direct Python translation of the C++ library by Ihsan Kehribar
# <ihsan@kehribar.me> and Omer Kilic <omerkilic@gmail.com>, version 0.9.
# It is released under the following license, the same as the original C++ source.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy of
# this software and associated documentation files (the "Software"), to deal in
# the Software without restriction, including without limitation the rights to
# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
# of the Software, and to permit persons to whom the Software is furnished to do
# so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import usb.core
import usb.control

# Constants--------------------------------------------------------------------

#USB constants
VENDOR_ID = 0x1781
PRODUCT_ID = 0x0C9F
USB_TIMEOUT = 5000
RX_BUFFER_SIZE = 64

#Pin Definitions

#General Purpose Pins
PIN1 = 1
PIN2 = 2
PIN3 = 5
PIN4 = 0

#ADC Channels
ADC_PIN3 = 0
ADC_PIN2 = 1
ADC_TEMP_SENS = 2

#PWM Pins
PWM1 = PIN4
PWM2 = PIN1

#Aliases
ADC0 = ADC_PIN3
ADC1 = ADC_PIN2
ADC2 = ADC_TEMP_SENS
PWMA = PWM1
PWMB = PWM2

#'AVR ISP' Pins
SCK_PIN = PIN2
MISO_PIN = PIN1
MOSI_PIN = PIN4
RESET_PIN = PIN3

#Servo constants
STEP_SIZE = 0.062   #in miliseconds
MIN_LIMIT = 0.8     #in miliseconds
MAX_LIMIT = 2.2     #in miliseconds
RANGE = 180         #in degrees

#Useful constants

INPUT = 1
OUTPUT = 0

AUTO_CS = 1
MANUAL_CS = 0

HIGH = 1
LOW = 0

# Code ------------------------------------------------------------------------

class device:
    """ Class to control a LittleWire USB Multi-Tool."""
    lw = None

    def __init__(self):
        """Finds the first littleWire and attaches to it"""
        self.lw = usb.core.find(idVendor=VENDOR_ID, idProduct=PRODUCT_ID)
        
        if self.lw == None:
            raise ValueError('Device not Found!')
        
        self.lw.set_configuration()
        
    def readFirmwareVersion(self):
        """Returns littleWire firmware version"""
        rtn = self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=34,
                                    wValue=0, wIndex=0,
                                    data_or_wLength=8, timeout=USB_TIMEOUT)
        version = rtn.pop()
        return str((version & 0xF0) >> 4 )+"." + str((version & 0x0F))
    
    def pinMode(self, pin, mode):
        """Sets GPIO pins to INPUT(1)or OUTPUT(0)"""
        if (mode == INPUT):
            self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=13,
                                  wValue=pin, wIndex=0,
                                  data_or_wLength=8, timeout=USB_TIMEOUT)
        else:
            self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=14,
                                  wValue=pin, wIndex=0,
                                  data_or_wLength=8, timeout=USB_TIMEOUT)
    
    def digitalWrite(self, pin, state):
        """Writes a digital HIGH (1) or LOW (0) to the selected GPIO"""
        if (state == 1):
            self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=18,
                                  wValue=pin, wIndex=0,
                                  data_or_wLength=8, timeout=USB_TIMEOUT)
        else:
            self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=19,
                                  wValue=pin, wIndex=0,
                                  data_or_wLength=8, timeout=USB_TIMEOUT)
        
    def digitalRead(self, pin):
        """Returns the digital status of the selected GPIO"""
        rtn = self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=20,
                                    wValue=pin, wIndex=0,
                                    data_or_wLength=8, timeout=USB_TIMEOUT)
        return rtn.pop()
        
    def analogread(self, channel):
        """Returns the analog value of the selected analog pin"""
        rtn = self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=15,
                                    wValue=channel, wIndex=0,
                                    data_or_wLength=8, timeout=USB_TIMEOUT)
        return rtn.pop()
    
    def pwm_init(self):
        """Initialises the PWM system.  Must be called before using PWM commands"""
        self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=16,
                              wValue=0, wIndex=0,
                              data_or_wLength=8, timeout=USB_TIMEOUT)
        
    def pwm_stop(self):
        """Stops the PWM outputs"""
        self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=32,
                              wValue=0, wIndex=0,
                              data_or_wLength=8, timeout=USB_TIMEOUT)

    def pwm_updateCompare(self, channelA=0, channelB=0):
        """Sets the PWM value for both channels"""
        self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=17,
                              wValue=channelA, wIndex=channelB,
                              data_or_wLength=8, timeout=USB_TIMEOUT)

    def pwm_updatePrescaler(self, value=1):
        """Sets the value of the PWM prescaler - controls the PWM frequency"""
        if (value == 1024):
            self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=22,
                                  wValue=4, wIndex=0,
                                  data_or_wLength=8, timeout=USB_TIMEOUT)
        elif (value == 256):
            self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=22,
                                  wValue=3, wIndex=0,
                                  data_or_wLength=8, timeout=USB_TIMEOUT)
        elif (value == 64):
            self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=22,
                                  wValue=2, wIndex=0,
                                  data_or_wLength=8, timeout=USB_TIMEOUT)
        elif (value == 8):
            self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=22,
                                  wValue=1, wIndex=0,
                                  data_or_wLength=8, timeout=USB_TIMEOUT)
        elif (value == 1):
            self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=22,
                                  wValue=0, wIndex=0,
                                  data_or_wLength=8, timeout=USB_TIMEOUT)
                
    def spi_init(self):
        """Initialises the SPI comms system"""
        self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=23,
                              wValue=0, wIndex=0,
                              data_or_wLength=8, timeout=USB_TIMEOUT)
        
    def spi_sendMessage(self, message):
        """Sends a message out on the SPI bus"""
        self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=21,
                              wValue=message, wIndex=0,
                              data_or_wLength=8, timeout=USB_TIMEOUT)

    def spi_sendMessageMulti(self, sendBuffer, length, mode):
        """Sends up to 4 bytes out on the SPI bus and reads the same number back"""
        if (length > 4):
            length = 4

        rtn = self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=(0xF0 + length + (mode <<3)),
                              wValue=(sendbuffer[1]<<8)+sendbuffer[0],
                              wIndex=(sendbuffer[3]<<8)+sendbuffer[2],
                              data_or_wLength=8, timeout=USB_TIMEOUT)
        return rtn
    
    def spi_updateDelay(self, duration):
        self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=31,
                              wValue=duration, wIndex=0,
                              data_or_wLength=8, timeout=USB_TIMEOUT)      

    def i2c_init(self):
        """Initialises the I2C comms system"""
        self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=24,
                              wValue=0, wIndex=0,
                              data_or_wLength=8, timeout=USB_TIMEOUT)
        
    def i2c_beginTransmission(self, address):
        self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=25,
                              wValue=address, wIndex=0,
                              data_or_wLength=8, timeout=USB_TIMEOUT)

    def i2c_send(self, message):
        self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=26,
                              wValue=message, wIndex=0,
                              data_or_wLength=8, timeout=USB_TIMEOUT)
        
    def i2c_endTransmission(self):
        self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=27,
                              wValue=0, wIndex=0,
                              data_or_wLength=8, timeout=USB_TIMEOUT)
        
    def i2c_requestFrom(self, address, numBytes, responseBuffer):
        rtn = self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=30,
                              wValue=address, wIndex=numBytes,
                              data_or_wLength=8, timeout=USB_TIMEOUT)
        return rtn

    def servo_init(self):
        """Initialises the Servo control"""
        self.pwm_init() #initialise the PWM hardware
        self.pinMode(PWMA, OUTPUT)
        self.pinMode(PWMB, OUTPUT)  #set the PWM pins output
        self.pwm_updatePrescaler(1024)  #make sure the PWM prescaler is set correctly

    def servo_updateLocation(self, locationChannelA, locationChannelB):
        """Moves servos to a new location"""
        locationChannelA = int((((locationChannelA/RANGE)*(MAX_LIMIT-MIN_LIMIT))+MIN_LIMIT)/STEP_SIZE)
        locationChannelB = int((((locationChannelB/RANGE)*(MAX_LIMIT-MIN_LIMIT))+MIN_LIMIT)/STEP_SIZE)
        self.pwm_updateCompare(locationChannelA, locationChannelB)

    def ws2812_write(self, pin, r, g, b):
        self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=54,
                              wValue=int((g<<8) | pin | 0x30), wIndex=int(b<<8),
                              data_or_wLength=8, timeout=USB_TIMEOUT)
	#lwStatus=usb_control_msg(lwHandle, 0xC0, 54, (g<<8) | pin | 0x30, (b<<8) | r, rxBuffer, 8, USB_TIMEOUT);
    
    def ws2812_flush(self, pin):
        self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=54,
                              wValue=int( pin | 0x10), wIndex=0,
                              data_or_wLength=8, timeout=USB_TIMEOUT)
	#lwStatus=usb_control_msg(lwHandle, 0xC0, 54, pin | 0x10, 0, rxBuffer, 8, USB_TIMEOUT);

    def ws2812_preload(self, r, g, b):
        self.lw.ctrl_transfer(bmRequestType=0xC0, bRequest=54,
                              wValue=int((g<<8) | 0x20), wIndex=(int(b<<8) | r),
                              data_or_wLength=8, timeout=USB_TIMEOUT)
	#lwStatus=usb_control_msg(lwHandle, 0xC0, 54, (g<<8) | 0x20, (b<<8) | r, rxBuffer, 8, USB_TIMEOUT);